test/integration/daemonset/daemonset_test.go

```diff
@@ -376,17 +144,77 @@ func TestInsufficientCapacityNodeDaemonDoesNotLaunchPod(t *testing.T) {
 		ds.Spec.UpdateStrategy = *strategy
 		_, err := dsClient.Create(ds)
 		if err != nil {
-			t.Fatalf("Failed to create DaemonSet: %v", err)
+			t.Fatalf("failed to create DaemonSet: %v", err)
 		}
 		node := newNode("node-with-limited-memory", nil)
 		node.Status.Allocatable = allocatableResources("100M", "200m")
 		_, err = nodeClient.Create(node)
 		if err != nil {
-			t.Fatalf("Failed to create node: %v", err)
+			t.Fatalf("failed to create node: %v", err)
 		}
 
 		validateFailedPlacementEvent(eventClient, t)
 
 		close(stopCh)
 	}
 }
+
+// A RollingUpdate DaemonSet should adopt existing pods when it is created
+func TestRollingUpdateDaemonSetExistingPodAdoption(t *testing.T) {
+	server, closeFn, dc, informers, clientset := setup(t)
+	defer closeFn()
+
+	tearDownFn := setupGC(t, server, 5)
+	defer tearDownFn()
+
+	ns := framework.CreateTestingNamespace("rolling-update-daemonset-existing-pod-adoption-test", server, t)
+	defer framework.DeleteTestingNamespace(ns, server, t)
+
+	dsClient := clientset.AppsV1().DaemonSets(ns.Name)
+	podClient := clientset.CoreV1().Pods(ns.Name)
+	nodeClient := clientset.CoreV1().Nodes()
+	podInformer := informers.Core().V1().Pods().Informer()
+	controllerRevisionClient := clientset.AppsV1().ControllerRevisions(ns.Name)
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	informers.Start(stopCh)
+	go dc.Run(5, stopCh)
+
+	// Step 1: create a RollingUpdate DaemonSet
+	dsName := "daemonset"
+	ds := newDaemonSet(dsName, ns.Name)
+	ds.Spec.UpdateStrategy = *newRollbackStrategy()
+	ds, err := dsClient.Create(ds)
+	if err != nil {
+		t.Fatalf("failed to create daemonset %q: %v", dsName, err)
+	}
+
+	nodeName := "single-node"
+	node := newNode(nodeName, nil)
+	_, err = nodeClient.Create(node)
+	if err != nil {
+		t.Fatalf("failed to create node %q: %v", nodeName, err)
+	}
+
+	// Validate everything works correctly (include marking daemon pod as ready)
+	validateDaemonSetPodsAndMarkReady(podClient, podInformer, 1, t)
+	validateDaemonSetStatus(dsClient, ds.Name, ds.Namespace, 1, t)
+
+	// Step 2: delete daemonset and orphan its pods
+	deleteDaemonSetAndOrphanPods(dsClient, podClient, controllerRevisionClient, podInformer, ds, t)
+
+	// Step 3: create 2rd daemonset to adopt the pods (no restart) as long as template matches
+	dsName2 := "daemonset-adopt-template-matches"
+	ds2 := newDaemonSet(dsName2, ns.Name)
+	ds2.Spec.UpdateStrategy = *newRollbackStrategy()
+	ds2, err = dsClient.Create(ds2)
+	if err != nil {
+		t.Fatalf("failed to create daemonset %q: %v", dsName2, err)
+	}
+	if !apiequality.Semantic.DeepEqual(ds2.Spec.Template, ds.Spec.Template) {
+		t.Fatalf(".spec.template of new daemonset %q and old daemonset %q are not the same", dsName2, dsName)
+	}
+
+	// Wait for pods and history to be adopted by 2nd daemonset
+	waitDaemonSetAdoption(podClient, controllerRevisionClient, podInformer, ds2, ds.Name, t)
```

Comment:
validate the daemonset status here using `validateDaemonSetStatus`?

Reply:
Yes, I will add another step to validate the DS's status.